### PR TITLE
Better optimization of consecutive keys in aggregation

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimized aggregation in some cases.

### Detailed description:
There are two optimizations in this PR:

1. Execute `addBatchSinglePlace` instead of `addBatch` in case when we have only one unique key in block.
2. Disable consecutive keys cache in case when miss rate is high.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 5630115faa66fac57e49246c05aa69b8780532bb with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57872/5630115faa66fac57e49246c05aa69b8780532bb/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7211143460">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57872/5630115faa66fac57e49246c05aa69b8780532bb/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by CurtizJ** [comment]
There are two optimizations in this PR:

1. Execute `addBatchSinglePlace` instead of `addBatch` in case when we have only one unique key in block.
2. Disable consecutive keys cache in case when miss rate is high.

**Comment by CurtizJ** [comment]
Some synthetic perf tests sped up a bit: 

Screenshot 2023-12-16 at 16 20 27
Screenshot 2023-12-16 at 16 20 51
Screenshot 2023-12-16 at 16 21 05

File: src/Common/ColumnsHashingImpl.h [hunk 1]
```diff
@@ -39,14 +50,19 @@ template <typename Value, bool consecutive_keys_optimization_>
...
 
-    bool check(const Value & value_) { return !empty && value == value_; }
+    bool check(const Value & value_) const { return value == value_; }
 
     template <typename Key>
-    bool check(const Key & key) { return !empty && value.first == key; }
+    bool check(const Key & key) const { return value.first == key; }
+
+    bool hasOnlyOneValue() const { return found && misses == 1; }
+    UInt64 getMisses() const { return misses; }
```
Comment (antonio2368) [inline src/Common/ColumnsHashingImpl.h:65]: what's the point of getter if fields are public and they don't have getters?

File: src/Core/Settings.h [hunk 2]
```diff
@@ -654,6 +654,7 @@ class IColumn;
     M(Bool, allow_aggregate_partitions_independently, false, "Enable independent aggregation of partitions on separate threads when partition key suits group by key. Beneficial when number of partitions close to number of cores and partitions have roughly the same size", 0) \
     M(Bool, force_aggregate_partitions_independently, false, "Force the use of optimization when it is applicable, but heuristics decided not to use it", 0) \
     M(UInt64, max_number_of_partitions_for_independent_aggregation, 128, "Maximal number of partitions in table to apply optimization", 0) \
+    M(Float, min_hit_rate_to_use_consecutive_keys_optimization, 0.5, "Minimal hit rate of a cache which is used for consecutive keys optimization in aggreagtion to keep it enabled", 0) \
```
Comment (antonio2368) [inline src/Core/Settings.h:657]:
```suggestion
    M(Float, min_hit_rate_to_use_consecutive_keys_optimization, 0.5, "Minimal hit rate of a cache which is used for consecutive keys optimization in aggregation to keep it enabled", 0) \
```

**Review by alexey-milovidov — APPROVED** [review]
This is amazing!

**Pull request merged**